web/modules/custom/custom_weather/src/Plugin/Block/CustomWeather.php

```diff
@@ -49,23 +45,24 @@ public static function create(ContainerInterface $container, array $configuratio
       $plugin_id,
       $plugin_definition,
       $container->get('config.factory'),
+      $container->get('Drupal\custom_weather\Service\UserCityHandler'),
     );
   }
 
   /**
    * {@inheritdoc}
    */
   public function build():array {
+    $city = $this->databaseService->getCityFromDatabase();
     $weatherData = $this->getWeatherApi();
     if ($weatherData) {
       $temp = round($weatherData['main']['temp'] - 273.15, 1);
       $weather_text = $weatherData['weather'][0]['main'];
-      $selected_city = $this->configFactory->get('custom_weather.settings')->get('selected_city');
       return [
         '#theme' => 'custom_weather_block',
         '#temp' => $temp,
         '#weather_text' => $weather_text,
-        '#selected_city' => $selected_city,
+        '#selected_city' => $city,
         '#cache' => [
           'max-age' => 30 * 60,
```

Comment:
What happens with that caching is
User 1 that have selected Kyiv opens the site = block cached
User 2 that have selected Lutsk opens the site = he sees weather for Kyiv, since it is cached